File: src/dotnet/QsoRipper.Engine.DotNet/Geodesy.cs
using System;
using QsoRipper.Domain;

namespace QsoRipper.Engine.DotNet;

/// <summary>
/// Spherical-Earth geodesy mirror of the Rust <c>qsoripper-core::geodesy</c>
/// module. Used by <see cref="ManagedGreatCircleGrpcService"/> to compute
/// distances, bearings, and great-circle samples for the
/// <c>GreatCircleService</c> RPC. Cross-engine results match within
/// floating-point tolerance.
/// </summary>
internal static class Geodesy
{
    public const double EarthRadiusKm = 6371.0088;
    public const uint DefaultSampleCount = 64;
    public const uint MinSampleCount = 2;
    public const uint MaxSampleCount = 512;

    /// <summary>
    /// Validate latitude/longitude. Throws <see cref="ArgumentException"/>
    /// for NaN, infinite, or out-of-range values.
    /// </summary>
    public static void ValidatePoint(GeoPoint point, string label)
    {
        ArgumentNullException.ThrowIfNull(point);
        if (double.IsNaN(point.Latitude) || double.IsInfinity(point.Latitude)
            || point.Latitude < -90.0 || point.Latitude > 90.0)
        {
            throw new ArgumentException(
                $"{label}: latitude {point.Latitude} is out of range or non-finite",
                nameof(point));
        }

        if (double.IsNaN(point.Longitude) || double.IsInfinity(point.Longitude)
            || point.Longitude < -180.0 || point.Longitude > 180.0)
        {
            throw new ArgumentException(
                $"{label}: longitude {point.Longitude} is out of range or non-finite",
                nameof(point));
        }
    }

    /// <summary>
    /// Resolve a Maidenhead locator (4, 6, or 8 chars; case-insensitive) to
    /// the center point of the locator's cell.
    /// </summary>
    public static GeoPoint MaidenheadToGeoPoint(string locator)
    {
        ArgumentNullException.ThrowIfNull(locator);
        var trimmed = locator.Trim().ToUpperInvariant();
        if (trimmed.Length is not (4 or 6 or 8))
        {
            throw new ArgumentException(
                $"invalid Maidenhead locator: expected 4, 6, or 8 characters, got {trimmed.Length}",
                nameof(locator));
        }

        int FieldDigit(char c, char lo, char hi, string desc)
        {
            if (c < lo || c > hi)
            {
                throw new ArgumentException($"invalid Maidenhead locator: bad {desc} '{c}'", nameof(locator));
            }
            return c - lo;
        }

        int Decimal(char c, string desc)
        {
            if (c < '0' || c > '9')
            {
                throw new ArgumentException($"invalid Maidenhead locator: bad {desc} '{c}'", nameof(locator));
            }
            return c - '0';
        }

        var fieldLon = FieldDigit(trimmed[0], 'A', 'R', "field");
        var fieldLat = FieldDigit(trimmed[1], 'A', 'R', "field");
        var lon = -180.0 + fieldLon * 20.0;
        var lat = -90.0 + fieldLat * 10.0;

        var sqLon = Decimal(trimmed[2], "square digit");
        var sqLat = Decimal(trimmed[3], "square digit");
        lon += sqLon * 2.0;
        lat += sqLat * 1.0;

        var lonStep = 2.0;
        var latStep = 1.0;

        if (trimmed.Length >= 6)
        {
            var subLon = FieldDigit(trimmed[4], 'A', 'X', "subsquare");
            var subLat = FieldDigit(trimmed[5], 'A', 'X', "subsquare");
            lonStep = 2.0 / 24.0;
            latStep = 1.0 / 24.0;
            lon += subLon * lonStep;
            lat += subLat * latStep;
        }

        if (trimmed.Length == 8)
        {
            var extLon = Decimal(trimmed[6], "extended digit");
            var extLat = Decimal(trimmed[7], "extended digit");
            lonStep /= 10.0;
            latStep /= 10.0;
            lon += extLon * lonStep;
            lat += extLat * latStep;
        }

        return new GeoPoint
        {
            Latitude = lat + (latStep / 2.0),
            Longitude = lon + (lonStep / 2.0),
        };
    }

    public static double DistanceKm(GeoPoint origin, GeoPoint target)
    {
        ArgumentNullException.ThrowIfNull(origin);
        ArgumentNullException.ThrowIfNull(target);
        var lat1 = ToRad(origin.Latitude);
        var lat2 = ToRad(target.Latitude);
        var dLat = ToRad(target.Latitude - origin.Latitude);
        var dLon = ToRad(target.Longitude - origin.Longitude);
        var a = (Math.Sin(dLat / 2) * Math.Sin(dLat / 2))
            + (Math.Cos(lat1) * Math.Cos(lat2) * Math.Sin(dLon / 2) * Math.Sin(dLon / 2));
        return 2.0 * EarthRadiusKm * Math.Atan2(Math.Sqrt(a), Math.Sqrt(1.0 - a));
    }

    /// <summary>
    /// Initial bearing in degrees (clockwise from true north). Returns
    /// <c>null</c> for coincident or antipodal endpoints where the bearing
    /// is undefined.
    /// </summary>
    public static double? InitialBearingDeg(GeoPoint origin, GeoPoint target)
    {
        ArgumentNullException.ThrowIfNull(origin);
        ArgumentNullException.ThrowIfNull(target);
        if (BearingUndefined(origin, target))
        {
            return null;
        }

        var lat1 = ToRad(origin.Latitude);
        var lat2 = ToRad(target.Latitude);
        var dLon = ToRad(target.Longitude - origin.Longitude);
        var y = Math.Sin(dLon) * Math.Cos(lat2);
        var x = (Math.Cos(lat1) * Math.Sin(lat2)) - (Math.Sin(lat1) * Math.Cos(lat2) * Math.Cos(dLon));
        return NormalizeBearing(ToDeg(Math.Atan2(y, x)));
    }

    public static double? FinalBearingDeg(GeoPoint origin, GeoPoint target)
    {
        var reverse = InitialBearingDeg(target, origin);
        return reverse.HasValue ? NormalizeBearing(reverse.Value + 180.0) : null;
    }

    /// <summary>
    /// Spherical linear interpolation along the great circle.
    /// Returns <paramref name="count"/> evenly spaced points (including endpoints).
    /// Caller must pass <paramref name="count"/> &gt;= 2.
    /// </summary>
    public static GeoPoint[] SampleGreatCircle(GeoPoint origin, GeoPoint target, uint count)
    {
        ArgumentNullException.ThrowIfNull(origin);
        ArgumentNullException.ThrowIfNull(target);
        var n = Math.Max(2u, count);
        var lat1 = ToRad(origin.Latitude);
        var lon1 = ToRad(origin.Longitude);
        var lat2 = ToRad(target.Latitude);
        var lon2 = ToRad(target.Longitude);
        var p1 = ToXyz(lat1, lon1);
        var p2 = ToXyz(lat2, lon2);

        var dot = Math.Clamp((p1.X * p2.X) + (p1.Y * p2.Y) + (p1.Z * p2.Z), -1.0, 1.0);
        var omega = Math.Acos(dot);
        var sinOmega = Math.Sin(omega);

        var samples = new GeoPoint[n];

        if (Math.Abs(sinOmega) < 1e-9)
        {
            for (var i = 0u; i < n; i++)
            {
                var t = (double)i / (n - 1);
                samples[i] = t < 0.5
                    ? new GeoPoint { Latitude = origin.Latitude, Longitude = origin.Longitude }
                    : new GeoPoint { Latitude = target.Latitude, Longitude = target.Longitude };
            }

            return samples;
        }

        for (var i = 0u; i < n; i++)
        {
            var t = (double)i / (n - 1);
            var a = Math.Sin((1.0 - t) * omega) / sinOmega;
            var b = Math.Sin(t * omega) / sinOmega;
            var x = (a * p1.X) + (b * p2.X);
            var y = (a * p1.Y) + (b * p2.Y);
            var z = (a * p1.Z) + (b * p2.Z);
            var lat = Math.Asin(z);
            var lon = Math.Atan2(y, x);
            samples[i] = new GeoPoint
            {
                Latitude = ToDeg(lat),
                Longitude = ToDeg(lon),
            };
        }

        return samples;
    }

    /// <summary>Resolve and validate the requested sample count.</summary>
    public static uint ResolveSampleCount(uint requested)
    {
        if (requested == 0)
        {
            return DefaultSampleCount;
        }
        if (requested < MinSampleCount || requested > MaxSampleCount)
        {
            throw new ArgumentOutOfRangeException(
                nameof(requested),
                requested,
                $"sample_count must be 0 (default) or in [{MinSampleCount}, {MaxSampleCount}]");
        }
        return requested;
    }

    private static bool BearingUndefined(GeoPoint origin, GeoPoint target)
    {
        var same = Math.Abs(origin.Latitude - target.Latitude) < 1e-9
            && Math.Abs(origin.Longitude - target.Longitude) < 1e-9;
        var antipodal = Math.Abs(origin.Latitude + target.Latitude) < 1e-9
            && Math.Abs(Math.Abs(origin.Longitude - target.Longitude) - 180.0) < 1e-9;
        return same || antipodal;
    }

    private static double NormalizeBearing(double deg)
    {
        var normalized = deg - (Math.Floor(deg / 360.0) * 360.0);
        return double.IsNaN(normalized) ? 0.0 : normalized;
    }

    private static (double X, double Y, double Z) ToXyz(double lat, double lon)
    {
        var cosLat = Math.Cos(lat);
        return (cosLat * Math.Cos(lon), cosLat * Math.Sin(lon), Math.Sin(lat));
    }

    private static double ToRad(double deg) => deg * Math.PI / 180.0;

    private static double ToDeg(double rad) => rad * 180.0 / Math.PI;
}

File: src/rust/qsoripper-core/src/geodesy.rs
//! Spherical-Earth geodesy: distance, bearing, and great-circle interpolation
//! used by the engine's `GreatCircleService`.
//!
//! All inputs and outputs are decimal degrees. Computations use the spherical
//! Earth model with mean radius `R = 6371.0088 km` (IUGG mean Earth radius).
//! This is sufficient for displaying contact distance/bearing on a UI map;
//! ellipsoidal precision (Vincenty/Karney) is not needed for those purposes.

use crate::proto::qsoripper::domain::GeoPoint;

/// Mean Earth radius in kilometres (IUGG/Wikipedia mean radius).
pub const EARTH_RADIUS_KM: f64 = 6371.0088;

/// Default number of geodesic samples returned when the caller asks for `0`.
pub const DEFAULT_SAMPLE_COUNT: u32 = 64;
/// Minimum supported sample count. A geodesic always includes both endpoints,
/// so the request must ask for at least 2 samples.
pub const MIN_SAMPLE_COUNT: u32 = 2;
/// Maximum supported sample count. Caps abuse without sacrificing visual
/// smoothness — 512 points is enough for a sub-pixel arc on a 4K display.
pub const MAX_SAMPLE_COUNT: u32 = 512;

/// Errors returned by the geodesy module. Mapped to gRPC `INVALID_ARGUMENT`
/// at the service boundary.
#[derive(Debug, thiserror::Error, PartialEq, Eq)]
pub enum GeodesyError {
    /// Latitude was outside `[-90, 90]` or non-finite.
    #[error("latitude {0} is out of range or non-finite")]
    LatitudeOutOfRange(String),
    /// Longitude was outside `[-180, 180]` or non-finite.
    #[error("longitude {0} is out of range or non-finite")]
    LongitudeOutOfRange(String),
    /// Sample count was below `MIN_SAMPLE_COUNT` or above `MAX_SAMPLE_COUNT`.
    #[error("sample_count {0} must be 0 (default) or in [{MIN_SAMPLE_COUNT}, {MAX_SAMPLE_COUNT}]")]
    SampleCountOutOfRange(u32),
    /// Maidenhead locator was empty, the wrong length, or contained invalid
    /// characters.
    #[error("invalid Maidenhead locator: {0}")]
    InvalidMaidenhead(String),
}

/// Validate a `GeoPoint`'s latitude and longitude.
///
/// # Errors
/// Returns `GeodesyError::LatitudeOutOfRange` or `LongitudeOutOfRange`
/// when the coordinates are non-finite or fall outside `[-90, 90]` /
/// `[-180, 180]`.
pub fn validate_point(point: &GeoPoint) -> Result<(), GeodesyError> {
    if !point.latitude.is_finite() || !(-90.0..=90.0).contains(&point.latitude) {
        return Err(GeodesyError::LatitudeOutOfRange(point.latitude.to_string()));
    }
    if !point.longitude.is_finite() || !(-180.0..=180.0).contains(&point.longitude) {
        return Err(GeodesyError::LongitudeOutOfRange(
            point.longitude.to_string(),
        ));
    }
    Ok(())
}

/// Resolve a Maidenhead locator (4, 6, or 8 characters) to its center
/// coordinates in decimal degrees. Locators are case-insensitive.
///
/// The classic Maidenhead system encodes longitude/latitude pairs as:
///   - field   (A..R)  → 20° × 10°
///   - square  (0..9)  → 2°  × 1°
///   - subsquare (a..x) → 5'  × 2.5'
///   - extended-subsquare (0..9) → 30" × 15"
///
/// # Errors
/// Returns `GeodesyError::InvalidMaidenhead` when the locator is the wrong
/// length or contains a character outside the expected range for its
/// position.
#[allow(clippy::indexing_slicing, clippy::similar_names)]
pub fn maidenhead_to_geopoint(locator: &str) -> Result<GeoPoint, GeodesyError> {
    let trimmed = locator.trim();
    let upper: Vec<char> = trimmed.to_ascii_uppercase().chars().collect();
    let len = upper.len();
    if !matches!(len, 4 | 6 | 8) {
        return Err(GeodesyError::InvalidMaidenhead(format!(
            "expected 4, 6, or 8 characters, got {len}"
        )));
    }
    // Field pair: A..R
    let field_lon = char_in_range(upper[0], 'A', 'R')
        .ok_or_else(|| GeodesyError::InvalidMaidenhead(format!("bad field char {}", upper[0])))?;
    let field_lat = char_in_range(upper[1], 'A', 'R')
        .ok_or_else(|| GeodesyError::InvalidMaidenhead(format!("bad field char {}", upper[1])))?;
    let mut lon = -180.0_f64 + f64::from(field_lon) * 20.0;
    let mut lat = -90.0_f64 + f64::from(field_lat) * 10.0;
    // Square pair: 0..9
    let sq_lon = digit(upper[2])
        .ok_or_else(|| GeodesyError::InvalidMaidenhead(format!("bad square digit {}", upper[2])))?;
    let sq_lat = digit(upper[3])
        .ok_or_else(|| GeodesyError::InvalidMaidenhead(format!("bad square digit {}", upper[3])))?;
    lon += f64::from(sq_lon) * 2.0;
    lat += f64::from(sq_lat) * 1.0;
    // Half cell width at this resolution (4-char locator = 2° lon × 1° lat).
    let mut lon_step = 2.0_f64;
    let mut lat_step = 1.0_f64;

    if len >= 6 {
        // Subsquare: A..X (need lowercase form for spec but we already uppered).
        let sub_lon = char_in_range(upper[4], 'A', 'X').ok_or_else(|| {
            GeodesyError::InvalidMaidenhead(format!("bad subsquare char {}", upper[4]))
        })?;
        let sub_lat = char_in_range(upper[5], 'A', 'X').ok_or_else(|| {
            GeodesyError::InvalidMaidenhead(format!("bad subsquare char {}", upper[5]))
        })?;
        lon_step = 2.0 / 24.0;
        lat_step = 1.0 / 24.0;
        lon += f64::from(sub_lon) * lon_step;
        lat += f64::from(sub_lat) * lat_step;
    }

    if len == 8 {
        let ext_lon = digit(upper[6]).ok_or_else(|| {
            GeodesyError::InvalidMaidenhead(format!("bad extended digit {}", upper[6]))
        })?;
        let ext_lat = digit(upper[7]).ok_or_else(|| {
            GeodesyError::InvalidMaidenhead(format!("bad extended digit {}", upper[7]))
        })?;
        lon_step /= 10.0;
        lat_step /= 10.0;
        lon += f64::from(ext_lon) * lon_step;
        lat += f64::from(ext_lat) * lat_step;
    }

    // Center of the cell: half a step in each direction.
    Ok(GeoPoint {
        latitude: lat + lat_step / 2.0,
        longitude: lon + lon_step / 2.0,
    })
}

fn char_in_range(c: char, lo: char, hi: char) -> Option<u32> {
    if (lo..=hi).contains(&c) {
        Some(u32::from(c) - u32::from(lo))
    } else {
        None
    }
}

fn digit(c: char) -> Option<u32> {
    c.to_digit(10)
}

/// Great-circle (Haversine) distance in kilometres between two points.
#[must_use]
pub fn distance_km(origin: &GeoPoint, target: &GeoPoint) -> f64 {
    let lat1 = origin.latitude.to_radians();
    let lat2 = target.latitude.to_radians();
    let dlat = (target.latitude - origin.latitude).to_radians();
    let dlon = (target.longitude - origin.longitude).to_radians();
    let a = (dlat / 2.0).sin().powi(2) + lat1.cos() * lat2.cos() * (dlon / 2.0).sin().powi(2);
    2.0 * EARTH_RADIUS_KM * a.sqrt().atan2((1.0 - a).sqrt())
}

/// Initial bearing in degrees from `origin` toward `target`, clockwise from
/// true north in `[0, 360)`. Returns `None` for coincident or antipodal
/// points where the bearing is undefined.
#[must_use]
pub fn initial_bearing_deg(origin: &GeoPoint, target: &GeoPoint) -> Option<f64> {
    if bearing_undefined(origin, target) {
        return None;
    }
    let lat1 = origin.latitude.to_radians();
    let lat2 = target.latitude.to_radians();
    let dlon = (target.longitude - origin.longitude).to_radians();
    let y = dlon.sin() * lat2.cos();
    let x = lat1.cos() * lat2.sin() - lat1.sin() * lat2.cos() * dlon.cos();
    Some(normalize_bearing(y.atan2(x).to_degrees()))
}

/// Final bearing in degrees at `target` of a great circle from `origin`.
/// Equivalent to the initial bearing of the reverse path, plus 180°.
#[must_use]
pub fn final_bearing_deg(origin: &GeoPoint, target: &GeoPoint) -> Option<f64> {
    initial_bearing_deg(target, origin).map(|b| normalize_bearing(b + 180.0))
}

fn bearing_undefined(origin: &GeoPoint, target: &GeoPoint) -> bool {
    // Same point => no direction.
    let same = (origin.latitude - target.latitude).abs() < 1e-9
        && (origin.longitude - target.longitude).abs() < 1e-9;
    // Antipodal: lat sum ≈ 0 AND |lon diff| ≈ 180.
    let antipodal = (origin.latitude + target.latitude).abs() < 1e-9
        && ((origin.longitude - target.longitude).abs() - 180.0).abs() < 1e-9;
    same || antipodal
}

fn normalize_bearing(deg: f64) -> f64 {
    let normalized = deg.rem_euclid(360.0);
    if normalized.is_nan() {
        0.0
    } else {
        normalized
    }
}

/// Spherical linear interpolation along the great circle from `origin`
/// to `target`. Returns `count` evenly spaced points (including endpoints).
/// `count` must be at least 2.
#[must_use]
#[allow(clippy::many_single_char_names, clippy::similar_names)]
pub fn sample_great_circle(origin: &GeoPoint, target: &GeoPoint, count: u32) -> Vec<GeoPoint> {
    let count = count.max(2);
    // Convert endpoints to unit-vector cartesian on the unit sphere.
    let (lat1, lon1) = (origin.latitude.to_radians(), origin.longitude.to_radians());
    let (lat2, lon2) = (target.latitude.to_radians(), target.longitude.to_radians());
    let p1 = lat_lon_to_xyz(lat1, lon1);
    let p2 = lat_lon_to_xyz(lat2, lon2);

    // Angular distance between the two points.
    let dot = (p1.0 * p2.0 + p1.1 * p2.1 + p1.2 * p2.2).clamp(-1.0, 1.0);
    let omega = dot.acos();
    let sin_omega = omega.sin();

    let mut out = Vec::with_capacity(count as usize);
    let last = f64::from(count - 1);

    // Degenerate cases: identical points or antipodal points => can't slerp.
    // Fall back to the endpoints repeated/linearly mixed; antipodal great
    // circles are non-unique so the rendered arc isn't meaningful anyway.
    if sin_omega.abs() < 1e-9 {
        for i in 0..count {
            let t = f64::from(i) / last;
            out.push(if t < 0.5 { *origin } else { *target });
        }
        return out;
    }

    for i in 0..count {
        let t = f64::from(i) / last;
        let a = ((1.0 - t) * omega).sin() / sin_omega;
        let b = (t * omega).sin() / sin_omega;
        let x = a * p1.0 + b * p2.0;
        let y = a * p1.1 + b * p2.1;
        let z = a * p1.2 + b * p2.2;
        let lat = z.asin();
        let lon = y.atan2(x);
        out.push(GeoPoint {
            latitude: lat.to_degrees(),
            longitude: lon.to_degrees(),
        });
    }
    out
}

fn lat_lon_to_xyz(lat: f64, lon: f64) -> (f64, f64, f64) {
    let cos_lat = lat.cos();
    (cos_lat * lon.cos(), cos_lat * lon.sin(), lat.sin())
}

/// Resolve and clamp the caller-requested sample count.
///
///   - 0          => `DEFAULT_SAMPLE_COUNT`
///   - 1          => error
///   - 2..=MAX    => honored
///   - > MAX      => error
///
/// # Errors
/// Returns `GeodesyError::SampleCountOutOfRange` when the request is `1`
/// or above `MAX_SAMPLE_COUNT`.
pub fn resolve_sample_count(requested: u32) -> Result<u32, GeodesyError> {
    if requested == 0 {
        Ok(DEFAULT_SAMPLE_COUNT)
    } else if (MIN_SAMPLE_COUNT..=MAX_SAMPLE_COUNT).contains(&requested) {
        Ok(requested)
    } else {
        Err(GeodesyError::SampleCountOutOfRange(requested))
    }
}

#[cfg(test)]
#[allow(
    clippy::expect_used,
    clippy::unwrap_used,
    clippy::indexing_slicing,
    clippy::similar_names,
    clippy::many_single_char_names
)]
mod tests {
    use super::*;

    fn pt(lat: f64, lon: f64) -> GeoPoint {
        GeoPoint {
            latitude: lat,
            longitude: lon,
        }
    }

    #[test]
    fn validate_accepts_normal_points() {
        assert!(validate_point(&pt(0.0, 0.0)).is_ok());
        assert!(validate_point(&pt(89.999, 179.999)).is_ok());
        assert!(validate_point(&pt(-90.0, -180.0)).is_ok());
    }

    #[test]
    fn validate_rejects_out_of_range_or_nan() {
        assert!(matches!(
            validate_point(&pt(91.0, 0.0)),
            Err(GeodesyError::LatitudeOutOfRange(_))
        ));
        assert!(matches!(
            validate_point(&pt(0.0, 181.0)),
            Err(GeodesyError::LongitudeOutOfRange(_))
        ));
        assert!(matches!(
            validate_point(&pt(f64::NAN, 0.0)),
            Err(GeodesyError::LatitudeOutOfRange(_))
        ));
        assert!(matches!(
            validate_point(&pt(0.0, f64::INFINITY)),
            Err(GeodesyError::LongitudeOutOfRange(_))
        ));
    }

    #[test]
    fn distance_seattle_to_london_is_about_7700_km() {
        // Seattle (KSEA) ≈ 47.45 N, -122.31; London (Heathrow) ≈ 51.47 N, -0.46.
        let d = distance_km(&pt(47.45, -122.31), &pt(51.47, -0.46));
        assert!((d - 7720.0).abs() < 30.0, "got {d}");
    }

    #[test]
    fn distance_paris_to_new_york_is_about_5837_km() {
        let d = distance_km(&pt(48.8566, 2.3522), &pt(40.7128, -74.0060));
        assert!((d - 5837.0).abs() < 20.0, "got {d}");
    }

    #[test]
    fn distance_to_self_is_zero() {
        let d = distance_km(&pt(34.0, -118.0), &pt(34.0, -118.0));
        assert!(d < 1e-6, "got {d}");
    }

    #[test]
    fn initial_bearing_known_pairs() {
        // Seattle → New York: roughly east-southeast (~78°).
        let b = initial_bearing_deg(&pt(47.45, -122.31), &pt(40.7128, -74.0060)).unwrap();
        assert!((b - 78.0).abs() < 5.0, "got {b}");

        // Paris → New York: roughly west-northwest (~291°).
        let b = initial_bearing_deg(&pt(48.8566, 2.3522), &pt(40.7128, -74.0060)).unwrap();
        assert!((b - 291.0).abs() < 5.0, "got {b}");
    }

    #[test]
    fn bearing_undefined_for_same_or_antipodal_points() {
        assert!(initial_bearing_deg(&pt(0.0, 0.0), &pt(0.0, 0.0)).is_none());
        assert!(initial_bearing_deg(&pt(10.0, 20.0), &pt(-10.0, -160.0)).is_none());
    }

    #[test]
    fn samples_endpoints_match() {
        let o = pt(47.45, -122.31);
        let t = pt(40.7128, -74.0060);
        let samples = sample_great_circle(&o, &t, 16);
        assert_eq!(samples.len(), 16);
        assert!((samples[0].latitude - o.latitude).abs() < 1e-6);
        assert!((samples[0].longitude - o.longitude).abs() < 1e-6);
        assert!((samples[15].latitude - t.latitude).abs() < 1e-6);
        assert!((samples[15].longitude - t.longitude).abs() < 1e-6);
    }

    #[test]
    fn samples_distance_matches_total() {
        let o = pt(47.45, -122.31);
        let t = pt(40.7128, -74.0060);
        let samples = sample_great_circle(&o, &t, 64);
        let mut sum = 0.0;
        for w in samples.windows(2) {
            sum += distance_km(&w[0], &w[1]);
        }
        let total = distance_km(&o, &t);
        // Polyline sum should match the geodesic distance to high precision.
        assert!(
            (sum - total).abs() / total < 1e-3,
            "sum={sum} total={total}"
        );
    }

    #[test]
    fn maidenhead_4char_center() {
        // CN87 = field C(2), N(13), square 8, 7. Cell spans lon -124..-122
        // and lat 47..48; the center is therefore (-123, 47.5).
        let p = maidenhead_to_geopoint("CN87").unwrap();
        assert!((p.longitude - (-123.0)).abs() < 1e-6, "lon={}", p.longitude);
        assert!((p.latitude - 47.5).abs() < 1e-6, "lat={}", p.latitude);
    }

    #[test]
    fn maidenhead_6char_center_in_4char_cell() {
        let p4 = maidenhead_to_geopoint("CN87").unwrap();
        let p6 = maidenhead_to_geopoint("CN87wn").unwrap();
        assert!((p6.longitude - p4.longitude).abs() < 2.0);
        assert!((p6.latitude - p4.latitude).abs() < 1.0);
    }

    #[test]
    fn maidenhead_case_insensitive() {
        let upper = maidenhead_to_geopoint("CN87WN").unwrap();
        let lower = maidenhead_to_geopoint("cn87wn").unwrap();
        assert!((upper.latitude - lower.latitude).abs() < 1e-9);
        assert!((upper.longitude - lower.longitude).abs() < 1e-9);
    }

    #[test]
    fn maidenhead_8char_center_in_6char_cell() {
        let p6 = maidenhead_to_geopoint("CN87wn").unwrap();
        let p8 = maidenhead_to_geopoint("CN87wn46").unwrap();
        assert!((p8.longitude - p6.longitude).abs() < (2.0 / 24.0));
        assert!((p8.latitude - p6.latitude).abs() < (1.0 / 24.0));
    }

    #[test]
    fn maidenhead_rejects_bad_input() {
        assert!(matches!(
            maidenhead_to_geopoint(""),
            Err(GeodesyError::InvalidMaidenhead(_))
        ));
        assert!(matches!(
            maidenhead_to_geopoint("CN8"),
            Err(GeodesyError::InvalidMaidenhead(_))
        ));
        assert!(matches!(
            maidenhead_to_geopoint("ZZ87"),
            Err(GeodesyError::InvalidMaidenhead(_))
        ));
        assert!(matches!(
            maidenhead_to_geopoint("CNAB"),
            Err(GeodesyError::InvalidMaidenhead(_))
        ));
    }

    #[test]
    fn resolve_sample_count_behaviour() {
        assert_eq!(resolve_sample_count(0).unwrap(), DEFAULT_SAMPLE_COUNT);
        assert_eq!(resolve_sample_count(2).unwrap(), 2);
        assert_eq!(resolve_sample_count(64).unwrap(), 64);
        assert_eq!(
            resolve_sample_count(MAX_SAMPLE_COUNT).unwrap(),
            MAX_SAMPLE_COUNT
        );
        assert!(matches!(
            resolve_sample_count(1),
            Err(GeodesyError::SampleCountOutOfRange(1))
        ));
        assert!(matches!(
            resolve_sample_count(MAX_SAMPLE_COUNT + 1),
            Err(GeodesyError::SampleCountOutOfRange(_))
        ));
    }
}
